src/dbwrapper.cpp

```diff
@@ -97,46 +97,59 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
 }
 
-static leveldb::Options GetOptions(size_t nCacheSize)
+static void GetOptions(size_t nCacheSize, leveldb::Options* options)
 {
-    leveldb::Options options;
-    options.block_cache = leveldb::NewLRUCache(nCacheSize / 2);
-    options.write_buffer_size = nCacheSize / 4; // up to two write buffers may be held in memory simultaneously
-    options.filter_policy = leveldb::NewBloomFilterPolicy(10);
-    options.compression = leveldb::kNoCompression;
-    options.info_log = new CBitcoinLevelDBLogger();
+    options->block_cache = leveldb::NewLRUCache(nCacheSize / 2);
+    options->write_buffer_size = nCacheSize / 4; // up to two write buffers may be held in memory simultaneously
+    options->filter_policy = leveldb::NewBloomFilterPolicy(10);
+    options->compression = leveldb::kNoCompression;
+    options->info_log = new CBitcoinLevelDBLogger();
     if (leveldb::kMajorVersion > 1 || (leveldb::kMajorVersion == 1 && leveldb::kMinorVersion >= 16)) {
         // LevelDB versions before 1.16 consider short writes to be corruption. Only trigger error
         // on corruption in later versions.
-        options.paranoid_checks = true;
+        options->paranoid_checks = true;
     }
-    SetMaxOpenFiles(&options);
-    return options;
+    SetMaxOpenFiles(options);
+}
+
+// Custom deleter for unique_ptr<leveldb::Options>.
+static void DeleteOptions(leveldb::Options* options)
+{
+    delete options->filter_policy;
+    options->filter_policy = nullptr;
+    delete options->info_log;
+    options->info_log = nullptr;
+    delete options->block_cache;
+    options->block_cache = nullptr;
+    options->env = nullptr;
 }
 
 CDBWrapper::CDBWrapper(const fs::path& path, size_t nCacheSize, bool fMemory, bool fWipe, bool obfuscate)
-    : m_name{path.stem().string()}
+    : options(new leveldb::Options, DeleteOptions), m_name{path.stem().string()}
 {
     penv = nullptr;
     readoptions.verify_checksums = true;
     iteroptions.verify_checksums = true;
     iteroptions.fill_cache = false;
     syncoptions.sync = true;
-    options = GetOptions(nCacheSize);
-    options.create_if_missing = true;
+    GetOptions(nCacheSize, options.get());
+    options->create_if_missing = true;
     if (fMemory) {
         penv = leveldb::NewMemEnv(leveldb::Env::Default());
-        options.env = penv;
+        options->env = penv;
     } else {
         if (fWipe) {
             LogPrintf("Wiping LevelDB in %s\n", path.string());
-            leveldb::Status result = leveldb::DestroyDB(path.string(), options);
+            leveldb::Status result = leveldb::DestroyDB(path.string(), *options);
             dbwrapper_private::HandleError(result);
         }
         TryCreateDirectories(path);
         LogPrintf("Opening LevelDB in %s\n", path.string());
     }
-    leveldb::Status status = leveldb::DB::Open(options, path.string(), &pdb);
+    leveldb::Status status = leveldb::DB::Open(*options, path.string(), &pdb);
+    if (!status.ok()) {
+        delete penv;
```

Comment:
If `penv` is deleted, then `~CDBWrapper()` gonna delete it again? Missing `= nullptr`?

Reply:
If `!status.ok` then `HandleError` throws and the destructor isn't called

Reply:
Maybe worth making `penv` a `unique_ptr` too, if we need manual handling with `delete`? Or do you think this would be kinda out of scope for this PR?

Adding addition `delete` into new code does not seem... "right" (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-newdelete).

Reply:
good point, I can look into making it a `unique_ptr`

Reply:
The latest change makes `penv` a unique_ptr